### ISSUE TITLE
`vim.loader` luac file paths can end up too long for windows to load

### ISSUE THREAD

**Issue by rktjmp** [neovim/neovim#25008]
### Problem

I hit this on Githubs CI, where I was setting `NVIM_APPNAME=$some_uuid` and generating long paths, not sure how applicable this is to "real world" use.

Windows paths are capped at 260 chars. The loader generates a rfc2396 encoded string of a module path, and puts that under the cache dir, so `c:\mod\file.lua` is stored at `c:\users\the-user\appdata\local\temp\<appname>\luac\c%3a%3fmod%3ffile.luac`.

If `mod\file.lua` is nested somewhere deep (say where a package manager might install it) and has some nesting of its own, and the user name is long, and/or the app_name is long, the encoded luac path can fail to write, and then the module fails to load, even if the module is loadable without the loader enabled.

There are ways to extend the path limit [here](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry), unsure if Neovim uses an application manifest though, and it requires a user registry edit.

In 0.9 the paths are just encoded as `%`, instead of `%FF`, so its possible this could be an unpleasant experience for some when upgrading in the future.

### Steps to reproduce

```powershell
# powershell -ExecutionPolicy Bypass x.ps1

$env:NVIM_APPNAME = [Guid]::NewGuid().ToString()

# create file in appname dir, with a long path
$appDataDir = [System.Environment]::GetFolderPath([System.Environment+SpecialFolder]::LocalApplicationData)
$luaDir = "$appDataDir\$env:NVIM_APPNAME\lua"

$long = "aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\aaa\a\a"
$modname = $long.Replace("\", ".")
mkdir "$luaDir\$long"
$luaFile = "$luaDir\$long.lua"
Set-Content -Path $luaFile -Value "print(1)"
echo $appDataDir
echo $luaFile

# install nvim 0.10 via msi package
& 'c:\Program Files\Neovim\bin\nvim' -v
# works, raw path is under limits
& 'c:\Program Files\Neovim\bin\nvim' $luaFile +":lua=require('$modname')"
# fails, luac path is too long.
# uses rfc2396 encoding, where \ becomes %f2 (1 char becomes 3)
& 'c:\Program Files\Neovim\bin\nvim' $luaFile +":lua vim.loader.enable()" +":lua=require('$modname')"

# install nvim 0.9 via choco
# uses plain \ to % encoding (1 char becomes 1)
nvim $luaFile +":lua=require('$modname')"
nvim $luaFile +":lua vim.loader.enable()" +":lua=require('$modname')"


```

### Expected behavior

The module loads, or maybe a nicer warning is given telling the user to swap OS or disable the loader for long paths automatically?

### Neovim version (nvim -v)

NVIM v0.10.0-dev-1025+g0e11bf0e1

### Vim (not Nvim) behaves the same?

NA

### Operating system/version

Windows 11

### Terminal name/version

Windows Terminal

### $TERM environment variable

NA?

### Installation

"latest" release msi package for

**Comment by clason** [comment]
One workaround is to simply disable the cache for this scenario.

**Comment by justinmk** [comment]
Yes, the cache is still considered experimental and it's a hyper-optimization, disabling it won't be noticeable for most users.

    vim.loader.disable()

**Comment by pappasam** [comment]
I believe I'm getting the same (or similar?) error, but on Ubuntu. Have disabled the cache, but I figure I'd provide the following data in case it's useful:

Error (with https://github.com/JoosepAlviste/nvim-ts-context-commentstring):

```
Error detected while processing BufReadPost Autocommands for "*":
Error executing lua callback: .../installs/neovim/nightly/share/nvim/runtime/filetype.lua:30: Error executing lua: .../installs
/neovim/nightly/share/nvim/runtime/filetype.lua:31: BufReadPost Autocommands for "*"..FileType Autocommands for "*": Vim(append
):Error executing lua callback: vim/loader.lua:0: ENAMETOOLONG: name too long: /home/sroeca/.cache/nvim/luac/%2fhome%2fsroeca%2
f.config%2fnvim%2fpack%2fpackager%2fstart%2fnvim-ts-context-commentstring%2flua%2fts_context_commentstring%2fintegrations%2fvim
_commentary.luac
```

System / Neovim version:

```
NVIM v0.10.0-dev-6405fa4
Build type: RelWithDebInfo
LuaJIT 2.1.1694940244
Run "nvim -V1 -v" for more info

Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy
```

**Comment by rktjmp** [comment]
`NAME_MAX` is 255 on my boring x86_64 system. I wonder if yours is particularly low? The unix limit is "per component", not the total path, but even then, the full path is 193ch and the file component is only 163. I did check saving a 254/255 ch long filenames with the nightly release in case it was built with some funny define but it acted as expected there.

The luac paths could just be the regular paths with a prefix, eg: `/home/x/y/z.lua` becomes `/home/x/.cache/nvim/luac/home/x/y/z.luac` which would avoid the issues on both platforms. Drive names (`c:/`) on windows still need special marshalling.

Alternatively, would it be preferable to silently ignore caching specific paths if `fs_open` fails (currently `assert`'d)? Or perhaps only if the reason contains `ENAMETOOLONG`?

We could issue a warning but I think that would mostly just be annoying since the user cant really fix it, short of disabling the loader, and given the choice would probably opt to cache 99% of files instead of 0%.

I can write a PR if there is consensus on how to handle it.

**Comment by justinmk** [comment]
> would it be preferable to silently ignore caching specific paths if `fs_open` fails (currently `assert`'d)? Or perhaps only if the reason contains `ENAMETOOLONG`

I think that's too subtle, makes analyzing performance more complicated. 

Why don't we hash the file paths? It makes inspecting the cache contents less obvious, but we could write out a mapping file (analogous to `~/.local/state/nvim/trust` )

**Comment by lewis6991** [comment]
> I think that's too subtle, makes analyzing performance more complicated.

True, but this is pretty edge case and hashing isn't so straight forward. To mitigate this we could just report issues to the log file?

**Comment by justinmk** [comment]
> To mitigate this we could just report issues to the log file?

We should at least show one message per session (notify_once or whatever), and then the rest can silently go to the logs. 

More generally though, Nvim by default should not freak out when an error happens. Solving https://github.com/neovim/neovim/issues/1029 is the single, central reason that other editors like Helix or VSCode "feel better" in general.
